Adds toy example and solves conflicts in README.md

leoisl opened the pull request:
Toy example has been added. @mbhall88 , could you please run it yourself and check if you manage to generate the same output (see `toy_example/README.md`)?

Also solved the `README.md` conflicts.

Sorry for not branching this on my fork, I am having an obscure issue with it. I will start with a fresh fork soon, I just want to be sure I don't lose anything important that has not been merged yet (probably not, but it is good to check).

iqbal-lab commented:
Given both of you will be away, i suggest we get me and leah to review this, as we will need to hold the fort and answer emails from people who can't run pandora

mbhall88 commented on README.md:
```diff
@@ -26,18 +26,7 @@
-  of diverse sets of genomes
+Pandora is a tool for bacterial genome analysis using a pangenome reference graph (PanRG). It allows gene presence/absence detection and genotyping of SNPs, indels and longer variants in one or a number of samples. Pandora works with Illumina or Nanopore data. 
 
 The pangenome reference graph (PanRG) is a collection of 'floating'
```
```suggestion
The PanRG is a collection of 'floating'
```

mbhall88 commented on README.md:
```diff
@@ -46,25 +35,16 @@ https://github.com/rmcolq/make_prg for a pipeline which can construct
 these PRGs from a set of aligned sequence files.
```
```suggestion
these PanRGs from a set of aligned sequence files.
```

mbhall88 commented on README.md:
```diff
@@ -46,25 +35,16 @@ https://github.com/rmcolq/make_prg for a pipeline which can construct
-Soon, in a galaxy not so far away, it will allow:
-- discovery of new variation not in the PRG
+- Output inferred mosaic of reference sequences for loci (eg genes) from the PanRG which are present;
+- Output a VCF showing the variation found within these loci, with respect to any reference path in the PRG;
```
Maybe we need to distinguish between a PanRG and a PRG as I am certain this will confuse some people. I assume in the second point we are referring to a local PRG?

leoisl replied:
Yeah, that makes sense... these are not my changes, these are the conflicts I solved when doing the merge, but I did not read the README in details, I just solved the conflicts. I will add a commit after going through your points that will define what is a PRG (not defined here), and will fix all these PRG/PanRG mentions

mbhall88 commented on README.md:
```diff
@@ -46,25 +35,16 @@ https://github.com/rmcolq/make_prg for a pipeline which can construct
-- discovery of new variation not in the PRG
+- Output inferred mosaic of reference sequences for loci (eg genes) from the PanRG which are present;
+- Output a VCF showing the variation found within these loci, with respect to any reference path in the PRG;
+- Discovery of new variation not in the PRG.
```
```suggestion
- Discovery of new variation not in the PanRG.
```

mbhall88 commented on README.md:
```diff
@@ -46,25 +35,16 @@ https://github.com/rmcolq/make_prg for a pipeline which can construct
-  shown with respect to the most informative recombinant path in the
-  PRG.
+- Output a matrix showing inferred copy-number of each locus in each sample genome;
+- Output a multisample pangenome VCF showing how including genotype calls for each sample in each of the loci;
```
This sentence seems incomplete?

leoisl replied:
It does indeed, this was before in the README... will fix this

mbhall88 commented on README.md:
```diff
@@ -46,25 +35,16 @@ https://github.com/rmcolq/make_prg for a pipeline which can construct
+- Output one VCF per locus. Variation is shown with respect to the most informative recombinant path in the PRG (see our paper).
 
-Warning - this code is still in development.
+Warning - pandora is not yet a production-ready tool yet. 
```
```suggestion
> **Warning - `pandora` is not yet a production-ready tool.** 
```

mbhall88 commented on README.md:
````diff
@@ -88,6 +68,11 @@ genes present
 pandora map <panrg.fa> <reads.fq>
 ```
 
+## Hands on a toy example
````
```suggestion
## Hands-on toy example
```

mbhall88 commented on README.md:
```diff
@@ -88,6 +68,11 @@ genes present
+## Hands on a toy example
+
+You can test `pandora` on a toy example following [this link](toy_example).
+There is no need to have pandora installed, as it is run inside containers.
```
```suggestion
There is no need to have `pandora` installed, as it is run inside containers.
```

mbhall88 commented on toy_example/README.md:
```diff
@@ -0,0 +1,62 @@
+# Toy example
+
+Here we present a walkthrough of running `pandora` on a toy example. We run: 1) `pandora` without de novo discovery; 2) `pandora` workflow (https://github.com/iqbal-lab-org/pandora_workflow), which runs pandora with and without de novo discovery (see Figure 2 of Colquhoun et al. "Nucleotide-resolution bacterial pan-genomics with reference graphs"). Although method 2) run both modes of `pandora`,  ...
```
```suggestion
Here we present a walkthrough of running `pandora` on a toy example. We run: 1) `pandora` without de novo discovery; 2) [`pandora` workflow](https://github.com/iqbal-lab-org/pandora_workflow), which runs `pandora` with and without de novo discovery (see Figure 2 of Colquhoun et al. "Nucleotide-resolution bacterial pan-genomics with reference graphs"). Although method 2) runs both modes of `pandora`, it is much more involved than method 1), as the user needs to configure and run a `snakemake` pipeline. If no de novo discovery is required, method 1 is a lot simpler to run (two commands) as opposed to running the pipeline. For completeness, we show both methods.
```

mbhall88 replied:
I would either link to the preprint when we refer to it in the text, or just say "our paper"

leoisl replied:
Will add link

mbhall88 commented on toy_example/README.md:
```diff
@@ -0,0 +1,62 @@
+msas/ : contains the MSAs of the two genes we are using as toy example here, GC00006032 and GC00010897;
+prgs/toy_prg.fa : contains the PRGs of the two genes, GC00006032 and GC00010897. These are fairly simple PRGs. GC00006032 contains 4 variant sites, each representing a SNP, while GC00010897 contains 5;
+reads/ : contains perfect simulated reads from two toy samples. We simulated perfect reads, where one sample genotypes to one allele of the variant sites, while the other sample genotypes towards the other allele;
+pandora_workflow_data : contains other input and configuration files to run the pandora workflow;
```
```suggestion
pandora_workflow_data/ : contains other input and configuration files to run the pandora workflow;
```

mbhall88 commented on toy_example/README.md:
```diff
@@ -0,0 +1,62 @@
+
+### Dependencies
+
+* `singularity`
```
```suggestion
* [`singularity`](https://sylabs.io/)
```

mbhall88 commented on toy_example/README.md:
```diff
@@ -0,0 +1,62 @@
+* `singularity`
+
+### Running
+`cd toy_example && ./run_pandora_nodenovo.sh`
```
Could you put this in a shell code fence? I tried doing it in the suggestions but it screws it up due to the code fence escaping the suggestion

mbhall88 commented on toy_example/README.md:
```diff
@@ -0,0 +1,62 @@
+
+### Quick look at the output
+
+`pandora` output will be located at dir `output_toy_example_no_denovo`.
```
```suggestion
`pandora` output will be located in directory `output_toy_example_no_denovo`.
```

mbhall88 commented on toy_example/README.md:
```diff
@@ -0,0 +1,62 @@
+
+`pandora` workflow output will be located at dir `pandora_workflow/output_toy_example_workflow/illumina/100x/random/compare_(no|with)denovo_global_genotyping`.
+
+Files `pandora_workflow/output_toy_example_workflow/illumina/100x/random/compare_nodenovo_global_genotyping/pandora_multisample_genotyped_global.vcf` and `output_toy_example_no_denovo/pandora_multisample_genotyped.vcf` both represent running `pandora` without de novo discovery and should be very similar files (just differentiating on some header lines, and on some statistics, due to slightly diff ...
```
The super long parentheses don't need to be in parentheses...

mbhall88 commented on toy_example/run_pandora_nodenovo.sh:
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
+set -eu
+
+singularity pull docker://rmcolq/pandora:latest
+singularity exec ./pandora-latest.simg pandora index prgs/toy_prg.fa
+singularity exec ./pandora-latest.simg pandora compare --genotype -o output_toy_example_no_denovo prgs/toy_prg.fa reads/read_index.tsv
```
```suggestion
URI="docker://rmcolq/pandora:latest"
singularity exec "$URI" pandora index prgs/toy_prg.fa
singularity exec "$URI" pandora compare --genotype -o output_toy_example_no_denovo prgs/toy_prg.fa reads/read_index.tsv
```

mbhall88 replied:
Also, the final `compare` command it missing a value to the `-o` parameter

mbhall88 replied:
Additionally, the changes I have suggested are more adaptable to singularity version differences. For example, this script fails for me as when I `singularity pull`, the image is named `pandora_latest.sif` not `.simg`

leoisl replied:
> Also, the final `compare` command it missing a value to the `-o` parameter

Not sure I understand this: the value to the `-o` param is `output_toy_example_no_denovo`

leoisl replied:
Argh, I think I will keep doing `singularity pull`, but will check if it is pulled into a `.simg` or `.sif`. Doing `singularity exec` is all fine, I actually preferred it, if it did not take so long to create the container runtime (singularity pull takes ~1min to be done, and then the 2 execs take <1s. If we don't pull before, each exec take >1 min)...

leoisl replied:
Will do `singularity exec` as you mentioned, but reduce it to a single command

mbhall88 commented:
> Toy example has been added. @mbhall88 , could you please run it yourself and check if you manage to generate the same output (see toy_example/README.md)?

Doesn't work for me. See my comments

> Given both of you will be away, i suggest we get me and leah to review this, as we will need to hold the fort and answer emails from people who can't run pandora

You're both welcome to review anything at any time :)

leoisl commented on README.md:
```diff
@@ -26,45 +26,25 @@
-  contained this chunk differed in their gene sequence. Variation is
-  shown with respect to the most informative recombinant path in the
-  PRG.
+- Output a matrix showing inferred copy-number of each locus in each sample genome;
```
@iqbal-lab  @mbhall88 : just tagging you to inform that I am changing this from "copy-number" to "presence-absence" ?

leoisl commented:
Updated the READMEs and the running script.

> Doesn't work for me. See my comments

Could you please retry? It works now for me on `singularity 2.x` version on local and on `noah`. I am trying with `singularity 3.5.0` on `noah`, but it is annoyingly stuck converting the docker image to sif image it seems... will update in case of success

> You're both welcome to review anything at any time :)

I still cant add Leah as a reviewer in this PR though...



> Also, we should add a link/citation for the preprint at the top of the README?

Added in the first paragraph of the Introduction, is this fine? https://github.com/rmcolq/pandora/pull/243/commits/82be6cfbe836736908311f30f714b1d238c98db2

leoisl commented:
Update: it also worked on `noah` with singularity `3.5.0`, but it took annoying 20 minutes to convert and start the container... but it works:

```
$ ./run_pandora_nodenovo.sh 
INFO:    Converting OCI blobs to SIF format
INFO:    Starting build...
Getting image source signatures
Copying blob 171857c49d0f skipped: already exists
Copying blob 419640447d26 skipped: already exists
Copying blob 61e52f862619 skipped: already exists
Copying blob 8ca3b4fddba5 skipped: already exists
Copying blob 08e2d9706b2e skipped: already exists
Copying blob 5a39e7e143ec skipped: already exists
Copying blob fc3080173dc4 skipped: already exists
Copying blob 32977d317cb6 skipped: already exists
Copying blob 21649c2c601e skipped: already exists
Copying blob 522e9250eb8e skipped: already exists
Copying config 74dc63d9ee done
Writing manifest to image destination
Storing signatures
2020/11/18 13:45:14  info unpack layer: sha256:171857c49d0f5e2ebf623e6cb36a8bcad585ed0c2aa99c87a055df034c1e5848
2020/11/18 13:45:16  warn xattr{etc/gshadow} ignoring ENOTSUP on setxattr "user.rootlesscontainers"
2020/11/18 13:45:16  warn xattr{/hps/nobackup/research/zi/leandro/temp/rootfs-438be42b-29a4-11eb-868b-18c04d1e0acf/etc/gshadow} destination filesystem does not support xattrs, further warnings will be suppressed
2020/11/18 13:45:55  info unpack layer: sha256:419640447d267f068d2f84a093cb13a56ce77e130877f5b8bdb4294f4a90a84f
2020/11/18 13:45:55  info unpack layer: sha256:61e52f862619ab016d3bcfbd78e5c7aaaa1989b4c295e6dbcacddd2d7b93e1f5
2020/11/18 13:45:55  info unpack layer: sha256:8ca3b4fddba53d643c35c9d5ce92f7ab957efb84b029cb73fba07cdb2c6034ee
2020/11/18 13:45:56  warn xattr{etc/gshadow} ignoring ENOTSUP on setxattr "user.rootlesscontainers"
2020/11/18 13:45:56  warn xattr{/hps/nobackup/research/zi/leandro/temp/rootfs-438be42b-29a4-11eb-868b-18c04d1e0acf/etc/gshadow} destination filesystem does not support xattrs, further warnings will be suppressed
2020/11/18 13:48:51  info unpack layer: sha256:08e2d9706b2ea21a60430e171b7dd671014c48739f1afab693cd48be2cdfa437
2020/11/18 13:48:51  warn xattr{boost_1.62.0/INSTALL} ignoring ENOTSUP on setxattr "user.rootlesscontainers"
2020/11/18 13:48:51  warn xattr{/hps/nobackup/research/zi/leandro/temp/rootfs-438be42b-29a4-11eb-868b-18c04d1e0acf/boost_1.62.0/INSTALL} destination filesystem does not support xattrs, further warnings will be suppressed
2020/11/18 14:00:21  info unpack layer: sha256:5a39e7e143ec201aa974bbe0a7353ecd6fc9ee03c6dddca4648a92463831845e
2020/11/18 14:04:06  info unpack layer: sha256:fc3080173dc470368ed2402f2fb8163e70a9ac36ee8170044f078bb9ed8d4ec7
2020/11/18 14:04:06  info unpack layer: sha256:32977d317cb6b55f43e239a0166f3ba3fc226c59d98bf0e66612fce218053fc1
2020/11/18 14:04:10  info unpack layer: sha256:21649c2c601e307a1bbb474d97b463a58c5a1227261be2353c5a73eae974ad14
2020/11/18 14:04:10  info unpack layer: sha256:522e9250eb8ef611f0aa73bf147df8ae5a9451e96b930ccb654d8b10fed67aa3
2020/11/18 14:04:11  warn xattr{var/cache/apt/archives/partial} ignoring ENOTSUP on setxattr "user.rootlesscontainers"
2020/11/18 14:04:11  warn xattr{/hps/nobackup/research/zi/leandro/temp/rootfs-438be42b-29a4-11eb-868b-18c04d1e0acf/var/cache/apt/archives/partial} destination filesystem does not support xattrs, further warnings will be suppressed
INFO:    Creating SIF file...
[2020-11-18 14:06:12.320460] [0x00002afb33c68ec0] [info]    Indexing PRG...
[2020-11-18 14:06:12.320560] [0x00002afb33c68ec0] [info]    Sketch PRG GC00010897 which has 13 nodes (0% done)
...
```

mbhall88 commented:
> I still cant add Leah as a reviewer in this PR though...

She needs to be a collaborator on the repo to be able to request her review. i.e. she needs to be able to accept/reject/merge PRs.

> Update: it also worked on noah with singularity 3.5.0, but it took annoying 20 minutes to convert and start the container... but it works. This is some cluster slowdown issue, as Leah managed to reproduce results with singularity 3 in 2-3 mins.

I suspect this could be to do with the location of you singularity cache. Run `echo $SINGULARITY_CACHEDIR` and see where it is pointing. I have mine pointing to a directory in my NFS. I think the default is a systems-defined location. I solved this problem with Leah and/or Grace before too I think...

mbhall88 commented on toy_example/README.md:
````diff
@@ -0,0 +1,70 @@
+
+### Running
+```
+cd toy_example && ./run_pandora_workflow.sh
````
This works now

mbhall88 reviewed (changes requested):
One other thing - this isn't a requested change though - would be renaming this directory to 'example' rather than 'toy_example'. This is more inline with most other software, is less verbose, and conveys the exact same impression/meaning.

leoisl commented:
> She needs to be a collaborator on the repo to be able to request her review. i.e. she needs to be able to accept/reject/merge PRs.

Yep, I misinterpreted a Zam's message (I thought she was already added as a collab, but I think we waiting for Rachel to add it)


> I suspect this could be to do with the location of you singularity cache. Run `echo $SINGULARITY_CACHEDIR` and see where it is pointing. I have mine pointing to a directory in my NFS. I think the default is a systems-defined location. I solved this problem with Leah and/or Grace before too I think...

I have never changed the `$SINGULARITY_CACHEDIR`. On `noah`, it is a user folder in `/hps/nobackup2/`. On `yoda` it is a `nfs` user folder. But I think both singularity `2` and `3` reads/writes to the same cache dir, and on version 2, the container is built in 1 min, instead of 20 mins in version 3... Anyway, I am happy this works, and won't go down the hole of debugging this slow down in the cluster :p



> One other thing - this isn't a requested change though - would be renaming this directory to 'example' rather than 'toy_example'. This is more inline with most other software, is less verbose, and conveys the exact same impression/meaning.

Done in https://github.com/rmcolq/pandora/commit/9d484cff13d8e76657532722499455900a69efc2 . Although I changed the dir name, I am wondering if it is still fine to refer to this example a toy example, for e.g. in the [README](https://github.com/rmcolq/pandora/blob/toy_example/example/README.md) and in the output dirs, like `output_toy_example_no_denovo`... should I change all mentions also?

leoisl commented on README.md:
```diff
@@ -148,7 +132,7 @@ git clone --single-branch https://github.com/rmcolq/pandora.git --recursive
 mkdir -p build
 cd build
-cmake ..
+cmake -DCMAKE_BUILD_TYPE=ReleaseWithAsserts ..
```
This actually triggered me to make a deeper debug of this build type in our `CMakeLists.txt`, and I just realised that the `-O3` flag was not being passed on into `make`. So, in the end we have been simply using the normal build instead of a compiler-optimised build all the time. You can see more details about this in https://github.com/iqbal-lab/pandora1_paper/issues/245 . A nice side effect of code review! Just reverted this change on https://github.com/rmcolq/pandora/pull/243/commits/6aa846a40edc60f8b0b75ed3121c94456f20988b

mbhall88 commented:
> Done in 9d484cf . Although I changed the dir name, I am wondering if it is still fine to refer to this example a toy example, for e.g. in the README and in the output dirs, like output_toy_example_no_denovo... should I change all mentions also?

No it's fine to leave the README etc. as it. It was pretty much just the directory I was referring to.
